client2/arq.go

```diff
@@ -0,0 +1,241 @@
+}
+
+func (a *ARQ) resend(surbID *[sConstants.SURBIDLength]byte) {
+	a.log.Info("resend")
+
+	a.lock.Lock()
+	message, ok := a.surbIDMap[*surbID]
+	if ok {
+		delete(a.surbIDMap, *surbID)
+		newsurbID := &[sConstants.SURBIDLength]byte{}
+		_, err := rand.Reader.Read(newsurbID[:])
+		if err != nil {
+			panic(err)
+		}
+
+		pkt, k, rtt, err := a.sphinxComposerSender.ComposeSphinxPacket(&Request{
+			WithSURB:          true,
+			SURBID:            newsurbID,
+			DestinationIdHash: message.DestinationIdHash,
+			RecipientQueueID:  message.RecipientQueueID,
+			Payload:           message.Payload,
+			IsSendOp:          true,
+		})
+		if err != nil {
+			panic(err)
+		}
+
+		message.SURBID = newsurbID
+		message.Key = k
+		message.ReplyETA = rtt
+		message.SentAt = time.Now()
+		message.Retransmissions += 1
+
+		a.surbIDMap[*newsurbID] = message
+		priority := uint64(message.SentAt.Add(message.ReplyETA).Add(RoundTripTimeSlop).UnixNano())
+		a.timerQueue.Push(priority, surbID)
+
+		err = a.sphinxComposerSender.SendSphinxPacket(pkt)
+		if err != nil {
+			a.log.Errorf("gc sphinx composer failure: %s", err.Error())
+		}
+
+	} else {
+		a.log.Error("gc SURB ID not found")
+	}
+	a.lock.Unlock()
+}
+
+// Has checks if a given SURB ID exists.
+func (a *ARQ) Has(surbID *[sConstants.SURBIDLength]byte) bool {
+	a.log.Info("Has")
+
+	a.lock.RLock()
+	_, ok := a.surbIDMap[*surbID]
+	a.lock.RUnlock()
+	return ok
+}
+
+// HandleACK removes the map entry for the given SURB ID AND returns
+// the APP ID and SURB Key so that the reply and be decrypted and routed
+// to the correct application.
+func (a *ARQ) HandleAck(surbID *[sConstants.SURBIDLength]byte) (*replyDescriptor, error) {
+	a.lock.Lock()
+
+	m, ok := a.surbIDMap[*surbID]
+	if !ok {
+		a.log.Error("failed to find SURB ID in ARQ map")
+		return nil, errors.New("failed to find SURB ID in ARQ map")
+	}
+	delete(a.surbIDMap, *surbID)
+
+	a.lock.Unlock()
+
+	return &replyDescriptor{
+		ID:      m.ID,
+		appID:   m.AppID,
+		surbKey: m.Key,
+	}, nil
+}
+
+// Send sends a message asynchronously. Sometime later, perhaps a reply will be received.
+func (a *ARQ) Send(appid uint64, id *[MessageIDLength]byte, payload []byte, providerHash *[32]byte, queueID []byte) error {
+	a.log.Info("Send")
+
+	a.lock.Lock()
+	defer a.lock.Unlock()
+
+	surbID := &[sConstants.SURBIDLength]byte{}
+	_, err := rand.Reader.Read(surbID[:])
+	if err != nil {
+		panic(err)
+	}
+
+	pkt, k, rtt, err := a.sphinxComposerSender.ComposeSphinxPacket(&Request{
+		WithSURB:          true,
+		SURBID:            surbID,
+		DestinationIdHash: providerHash,
+		RecipientQueueID:  queueID,
+		Payload:           payload,
+		IsSendOp:          true,
+	})
+	if err != nil {
+		panic(err)
+	}
+
+	message := &ARQMessage{
+		AppID:    appid,
+		ID:       id,
+		SURBID:   surbID,
+		Payload:  payload,
+		Key:      k,
+		SentAt:   time.Now(),
+		ReplyETA: rtt,
+	}
+	a.surbIDMap[*surbID] = message
+	p := time.Duration(message.ReplyETA + RoundTripTimeSlop)
```

Comment:
beyond here it doesn't look like we need the lock anymore?

Reply:
fixed